File: wai_oa/wai_oa_control_center/picovoice_modules/pvspeaker/demo/python/pv_speaker_demo.py
#
# Copyright 2024 Picovoice Inc.
#
# You may not use this file except in compliance with the license. A copy of the license is located in the "LICENSE"
# file accompanying this source.
#
# Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
# an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
# specific language governing permissions and limitations under the License.
#


import argparse
import wave
import array

import threading

from pvspeaker import PvSpeaker


def blocking_call(speaker):
    speaker.flush()


def worker_function(speaker, completion_event):
    blocking_call(speaker)
    completion_event.set()


def split_list(input_list, x):
    return [input_list[i:i + x] for i in range(0, len(input_list), x)]


def main():
    parser = argparse.ArgumentParser()

    parser.add_argument(
        "--show_audio_devices",
        "-s",
        help="List of audio devices currently available for use.",
        action="store_true")

    parser.add_argument(
        "--audio_device_index",
        "-d",
        help="Index of input audio device.",
        type=int,
        default=-1)

    parser.add_argument(
        "--input_wav_path",
        "-i",
        help="Path to PCM WAV file to be played.",
        default=None)

    parser.add_argument(
        "--buffer_size_secs",
        "-b",
        help="Size of internal PCM buffer in seconds.",
        type=int,
        default=20)

    parser.add_argument(
        "--output_wav_path",
        "-o",
        help="Path to the output WAV file where the PCM data passed to PvSpeaker will be written.",
        default=None)

    args = parser.parse_args()

    if args.show_audio_devices:
        devices = PvSpeaker.get_available_devices()
        for i in range(len(devices)):
            print("index: %d, device name: %s" % (i, devices[i]))
    else:
        device_index = args.audio_device_index
        input_path = args.input_wav_path
        buffer_size_secs = args.buffer_size_secs
        output_path = args.output_wav_path

        wavfile = None
        speaker = None

        try:
            if input_path is not None:
                wavfile = wave.open(input_path, "rb")

                sample_rate = wavfile.getframerate()
                bits_per_sample = wavfile.getsampwidth() * 8
                num_channels = wavfile.getnchannels()
                num_samples = wavfile.getnframes()

                if bits_per_sample != 8 and bits_per_sample != 16 and bits_per_sample != 24 and bits_per_sample != 32:
                    print(f"Unsupported bits per sample: {bits_per_sample}")
                    wavfile.close()
                    exit()

                if num_channels != 1:
                    print("WAV file must have a single channel (MONO)")
                    wavfile.close()
                    exit()

                speaker = PvSpeaker(
                    sample_rate=sample_rate,
                    bits_per_sample=bits_per_sample,
                    buffer_size_secs=buffer_size_secs,
                    device_index=device_index)
                print("pvspeaker version: %s" % speaker.version)
                print("Using device: %s" % speaker.selected_device)

                wav_bytes = wavfile.readframes(num_samples)

                pcm = None
                if bits_per_sample == 8:
                    pcm = list(array.array('B', wav_bytes))
                elif bits_per_sample == 16:
                    pcm = list(array.array('h', wav_bytes))
                elif bits_per_sample == 24:
                    pcm = []
                    for i in range(0, len(wav_bytes), 3):
                        sample = int.from_bytes(wav_bytes[i:i + 3], byteorder='little', signed=True)
                        pcm.append(sample)
                elif bits_per_sample == 32:
                    pcm = list(array.array('i', wav_bytes))

                pcm_list = split_list(pcm, sample_rate)
                speaker.start()

                if output_path:
                    speaker.write_to_file(output_path)

                print("Playing audio...")
                for pcm_sublist in pcm_list:
                    sublist_length = len(pcm_sublist)
                    total_written_length = 0
                    while total_written_length < sublist_length:
                        written_length = speaker.write(pcm_sublist[total_written_length:])
                        total_written_length += written_length

                print("Waiting for audio to finish...")

                completion_event = threading.Event()
                worker_thread = threading.Thread(target=worker_function, args=(speaker, completion_event))
                worker_thread.start()
                completion_event.wait()
                worker_thread.join()

                speaker.stop()

                print("Finished playing audio...")
                wavfile.close()

        except KeyboardInterrupt:
            print("\nStopped...")
            speaker.stop()
        finally:
            if speaker is not None:
                print("Deleting PvSpeaker...")
                speaker.delete()
            if wavfile is not None:
                wavfile.close()


if __name__ == "__main__":
    main()

File: wai_oa/wai_oa_control_center/picovoice_modules/pvspeaker/demo/c/pv_speaker_demo.c
/*
    Copyright 2024 Picovoice Inc.

    You may not use this file except in compliance with the license. A copy of the license is located in the "LICENSE"
    file accompanying this source.

    Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
    an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
    specific language governing permissions and limitations under the License.
*/

#include <getopt.h>
#include <signal.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "pv_speaker.h"

static volatile bool is_interrupted = false;

pv_speaker_t *speaker = NULL;

void interrupt_handler(int _) {
    (void) _;
    is_interrupted = true;
    pv_speaker_status_t status = pv_speaker_stop(speaker);
    if (status != PV_SPEAKER_STATUS_SUCCESS) {
        fprintf(stderr, "Failed to stop device with %s.\n", pv_speaker_status_to_string(status));
        exit(1);
    }
    fprintf(stdout, "\nStopped...\n");
}

static struct option long_options[] = {
        {"show_audio_devices", no_argument,       NULL, 's'},
        {"input_wav_path",     required_argument, NULL, 'i'},
        {"audio_device_index", required_argument, NULL, 'd'},
        {"buffer_size_secs",   required_argument, NULL, 'b'},
        {"output_wav_path",    required_argument, NULL, 'o'}
};

static void print_usage(const char *program_name) {
    fprintf(stderr,
            "Usage : %s -i INPUT_WAV_PATH [-d AUDIO_DEVICE_INDEX] [-b BUFFER_SIZE_SECS]\n"
            "        %s --show_audio_devices\n",
            program_name,
            program_name);
}

static void show_audio_devices(void) {
    char **device_list = NULL;
    int32_t device_list_length = 0;

    // List devices
    pv_speaker_status_t status = pv_speaker_get_available_devices(
            &device_list_length,
            &device_list);
    if (status != PV_SPEAKER_STATUS_SUCCESS) {
        fprintf(stderr, "Failed to get audio devices with: %s.\n", pv_speaker_status_to_string(status));
        exit(1);
    }

    fprintf(stdout, "Printing devices...\n");
    for (int32_t i = 0; i < device_list_length; i++) {
        fprintf(stdout, "index: %d, name: %s\n", i, device_list[i]);
    }

    pv_speaker_free_available_devices(device_list_length, device_list);
}

typedef struct {
    uint8_t chunk_id[4];
    uint32_t chunk_size;
    uint8_t format[4];
    uint8_t subchunk1_id[4];
    uint32_t subchunk1_size;
    uint16_t audio_format;
    uint16_t num_channels;
    uint32_t sample_rate;
    uint32_t byte_rate;
    uint16_t block_align;
    uint16_t bits_per_sample;
    uint8_t subchunk2_id[4];
    uint32_t subchunk2_size;
} wav_header;

void *read_wav_file(const char *filename, uint32_t *num_samples, uint32_t *sample_rate, uint16_t *bits_per_sample) {
    FILE *file = fopen(filename, "rb");
    if (!file) {
        perror("Unable to open file");
        return NULL;
    }

    wav_header header;

    fread(&header, sizeof(header), 1, file);

    if (header.chunk_id[0] != 'R' || header.chunk_id[1] != 'I' || header.chunk_id[2] != 'F' || header.chunk_id[3] != 'F' ||
        header.format[0] != 'W' || header.format[1] != 'A' || header.format[2] != 'V' || header.format[3] != 'E') {
        fclose(file);
        fprintf(stderr, "Invalid WAV file\n");
        return NULL;
    }

    if (header.audio_format != 1) {
        fclose(file);
        fprintf(stderr, "WAV file format must be PCM type\n");
        return NULL;
    }

    if (header.num_channels != 1) {
        fclose(file);
        fprintf(stderr, "WAV file must have a single channel (MONO)\n");
        return NULL;
    }

    *sample_rate = header.sample_rate;
    *bits_per_sample = header.bits_per_sample;
    uint32_t bytes_per_sample = header.bits_per_sample / 8;
    *num_samples = header.subchunk2_size / bytes_per_sample;

    void *pcm_data = malloc(header.subchunk2_size);

    if (!pcm_data) {
        perror("Memory allocation failed");
        fclose(file);
        return NULL;
    }

    fread(pcm_data, header.subchunk2_size, 1, file);

    fclose(file);

    return pcm_data;
}

int main(int argc, char *argv[]) {
    const char *input_wav_path = NULL;
    int32_t device_index = -1;
    int32_t buffer_size_secs = 20;
    const char *output_wav_path = NULL;

    int c;
    while ((c = getopt_long(argc, argv, "si:d:b:o:", long_options, NULL)) != -1) {
        switch (c) {
            case 's':
                show_audio_devices();
                return 0;
            case 'i':
                input_wav_path = optarg;
                break;
            case 'd':
                device_index = (int32_t) strtol(optarg, NULL, 10);
                break;
            case 'b':
                buffer_size_secs = (int32_t) strtol(optarg, NULL, 10);
                break;
            case 'o':
                output_wav_path = optarg;
                break;
            default:
                exit(1);
        }
    }

    if (!input_wav_path) {
        print_usage(argv[0]);
        exit(1);
    }

    signal(SIGINT, interrupt_handler);
    fprintf(stdout, "pv_speaker version: %s\n", pv_speaker_version());

    uint32_t num_samples, sample_rate;
    uint16_t bits_per_sample;
    void *pcm_data = read_wav_file(input_wav_path, &num_samples, &sample_rate, &bits_per_sample);

    fprintf(stdout, "Initializing pv_speaker...\n");
    pv_speaker_status_t status = pv_speaker_init(
            sample_rate,
            bits_per_sample,
            buffer_size_secs,
            device_index,
            &speaker);
    if (status != PV_SPEAKER_STATUS_SUCCESS) {
        fprintf(stderr, "Failed to initialize device with %s.\n", pv_speaker_status_to_string(status));
        exit(1);
    }

    const char *selected_device = pv_speaker_get_selected_device(speaker);
    fprintf(stdout, "Selected device: %s.\n", selected_device);

    if (output_wav_path != NULL) {
        pv_speaker_write_to_file(speaker, output_wav_path);
    }

    status = pv_speaker_start(speaker);
    if (status != PV_SPEAKER_STATUS_SUCCESS) {
        fprintf(stderr, "Failed to start device with %s.\n", pv_speaker_status_to_string(status));
        exit(1);
    }

    fprintf(stdout, "Playing audio...\n");
    if (pcm_data) {
        int8_t *pcm = (int8_t *) pcm_data;
        int32_t total_written_length = 0;

        while (!is_interrupted && total_written_length < num_samples) {
            int32_t written_length = 0;
            status = pv_speaker_write(
                    speaker,
                    &pcm[total_written_length * bits_per_sample / 8],
                    num_samples - total_written_length,
                    &written_length);
            if (status != PV_SPEAKER_STATUS_SUCCESS) {
                fprintf(stderr, "Failed to write with %s.\n", pv_speaker_status_to_string(status));
                exit(1);
            }
            total_written_length += written_length;
        }

        free(pcm);
    }

    if (!is_interrupted) {
        fprintf(stdout, "Waiting for audio to finish...\n");
        int8_t *pcm = NULL;
        int32_t pcm_length = 0;
        int32_t written_length = 0;
        status = pv_speaker_flush(speaker, pcm, pcm_length, &written_length);
        if (status != PV_SPEAKER_STATUS_SUCCESS) {
            fprintf(stderr, "Failed to flush pcm with %s.\n", pv_speaker_status_to_string(status));
            exit(1);
        }
    }

    if (!is_interrupted) {
        fprintf(stdout, "Finished playing audio...\n");
        status = pv_speaker_stop(speaker);
        if (status != PV_SPEAKER_STATUS_SUCCESS) {
            fprintf(stderr, "Failed to stop device with %s.\n", pv_speaker_status_to_string(status));
            exit(1);
        }
    }

    fprintf(stdout, "Deleting pv_speaker...\n");
    pv_speaker_delete(speaker);

    return 0;
}
